capacity4more/modules/c4m/features_og/c4m_features_og_events/c4m_features_og_events.module

```diff
@@ -219,38 +219,110 @@ function c4m_features_og_events_entity_info_alter(&$entity_info) {
-  $range = $user->uid ? 3 : 1;
+  if ($user->uid) {
+    // Check if content filter is applied.
+    $user_wrapper = entity_metadata_wrapper('user', $user->uid);
+    $interests = $user_wrapper->c4m_related_topic->raw();
+    $filter = !empty($_GET['filter']) ? $_GET['filter'] : '';
+    $groups = array();
+    $range = 3;
+  }
+  else {
+    $groups = c4m_og_get_groups_by_access_type('public');
+    $range = 1;
+  }
+
+  // Filter: Case "interests": Public groups and restricted groups which user's mail address is matching the domain AND tagged with one or more of the users topics of interest.
+  // Filter: Case "groups": User's groups, If there's no groups, No events are shown.
+  // Filter: Default: Show all events from all groups.
+  switch ($filter) {
+    case 'interests':
+
+      $groups = og_get_groups_by_user();
+      $groups = !empty($groups['node']) ? $groups['node'] : array();
+
+      if (!empty($interests)) {
+        $query = new EntityFieldQuery();
+
+        $query
+          ->entityCondition('entity_type', 'node')
+          ->entityCondition('bundle', 'group')
+          ->propertyCondition('status', NODE_PUBLISHED)
+          ->fieldCondition('c4m_related_topic', 'target_id', $interests, 'IN');
+        if (!empty($groups)) {
+          $query->propertyCondition('nid', $groups, 'NOT IN');
+        }
+        $result = $query->execute();
+
+        $groups += !empty($result['node']) ? array_keys($result['node']) : array();
+      }
+      break;
+    case 'groups':
+      $groups = og_get_groups_by_user();
+      $groups = !empty($groups['node']) ? $groups['node'] : array('no_groups');
+      break;
+    default:
+      break;
+  }
 
   $query = new EntityFieldQuery();
-  $result = $query
+  $query
     ->entityCondition('entity_type', 'node')
     ->entityCondition('bundle', 'event')
     ->propertyCondition('status', NODE_PUBLISHED)
-    ->fieldCondition(OG_AUDIENCE_FIELD, 'target_id', isset($groups['node']) ? $groups['node'] : $groups, 'IN')
-    ->fieldCondition('c4m_datetime_end', 'value', $date, '>=')
+    ->fieldCondition('c4m_datetime_end', 'value', $date, '>=');
+
+  if (!empty($groups)) {
+    $query->fieldCondition(OG_AUDIENCE_FIELD, 'target_id', $groups, 'IN');
+  }
+
+  $result = $query
     ->fieldOrderBy('c4m_datetime_end', 'value', 'ASC')
     ->range(0, $range)
     ->execute();
 
-  if (!empty($result['node'])) {
-    foreach ($result['node'] as $event) {
-      $node = node_view(node_load($event->nid), 'upcoming_event');
-      $events .= drupal_render($node);
-    }
+  if (empty($result['node'])) {
+    return FALSE;
   }
 
+  // Render title.
+  $title['element'] = array(
+    '#tag'   => 'h4',
+    '#value' => t('Upcoming events'),
+  );
+  $events .= theme_html_tag($title);
+
+  // Render events.
+  foreach ($result['node'] as $event) {
+    $node = node_view(node_load($event->nid), 'upcoming_event');
+    $events .= drupal_render($node);
+  }
+
+  // Render "see more" link.
```

Comment:
I think it's event better to put the rendering of the output in a template.
Then you don't have to add the title & link in render arrays…

And gives the advantage we can override the template later if necessary.

We should avoid 1 function doing to much.
We should put rendering of output in theme functions and/or templates.